Персептрон - это простейшая модель нейрона, предложенная Фрэнком Розенблаттом в 1958 году. В задаче классификации ирисов (2 класса) он выступает в роли линейного классификатора, который пытается провести прямую линию (гиперплоскость), разделяющую объекты разных классов.

Нейрон принимает на вход вектор признаков $x = (x_1, x_2, ..., x_n)$. Каждому признаку соответствует свой вес $w_i$, который определяет важность этого признака для принятия решения.

Процесс вычисления состоит из двух этапов:

1. Линейная комбинация (Взвешенная сумма): $$z = \sum_{i=1}^{n} w_i x_i + b$$где $b$ (bias) — это смещение, позволяющее сдвигать разделяющую прямую относительно начала координат.

2. Функция активации: В классическом персептроне используется пороговая функция:$$y = \text{step}(z) = \begin{cases} 1, & \text{если } z \geq 0 \\ 0, & \text{если } z < 0 \end{cases}$$

Обучение заключается в корректировке весов $w$ при каждой ошибке. Если предсказание $\hat{y}$ не совпало с истинным значением $y$, веса обновляются по формуле:

$$\Delta w_i = \eta \cdot (y - \hat{y}) \cdot x_i$$
$$w_{new} = w_{old} + \Delta w_i$$

Где $\eta$ - это скорость обучения (learning rate), определяющая размер шага при обновлении весов.

Персептрон гарантированно сходится (находит решение) только в том случае, если данные линейно разделимы.

In [ ]:
import numpy as np
from sklearn.datasets import load_iris
from sklearn.linear_model import Perceptron
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler

# 1. Загрузка данных
iris = load_iris()
X = iris.data
y = iris.target

# 2. Оставляем только два класса (0: setosa, 1: versicolor) 
# и два признака (0: sepal length, 1: sepal width)
mask = y < 2
X_filtered = X[mask, :2]
y_filtered = y[mask]

# 3. Стандартизация признаков
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X_filtered)

# 4. Инициализация и обучение модели
clf = Perceptron(random_state=42, max_iter=1000)
clf.fit(X_scaled, y_filtered)

# 5. Получение предсказаний
y_pred = clf.predict(X_scaled)

# 6. Вывод результатов
print("--- Результаты обучения ---")
print(f"Веса (Weights): {clf.coef_}")
print(f"Смещение (Bias): {clf.intercept_}")
print(f"Точность (Accuracy): {accuracy_score(y_filtered, y_pred) * 100:.2f}%")

--- Результаты обучения со стандартизацией ---
Веса (Weights): [[ 2.5059446  -1.88941004]]
Смещение (Bias): [0.]
Точность (Accuracy): 100.00%


Веса ($w_1, w_2$): Показывают вклад длины и ширины чашелистика в итоговый результат.

- 2.5059446 (Sepal Length): Положительный вес. Чем длиннее чашелистик, тем больше вероятность, что это класс 1 (Versicolor).

- -1.88941004 (Sepal Width): Отрицательный вес. Чем шире чашелистик, тем больше вероятность, что это класс 0 (Setosa).

Смещение ($b$): Определяет положение границы решения.

Точность (Accuracy): Доля правильных ответов.